### PR TITLE
Improve diagnostic output during deployment

### PR DESCRIPTION
**Pull request by jmcphers**
This change addresses feedback from the support team concerning error attribution in deployment failures. Deployment is a complex process involving many parts (including Packrat for bundling, lots of logic in the rsconnect package itself, and then unbundling/rendering on the server end), and when a failure occurs it's not clear exactly where it's coming from. 

Towards that end, this change bundles a number of improvements:

There is now a **verbose output mode** for deployment. In this mode, the log is annotated with extra information, including the `deployApp` call and R session information, and most tracing options are enabled for the duration of the deployment, including those that log more detailed HTTP errors, and the JSON communication between the client and server.  Note that this part of the change changes function signatures. 

There is a **new option**, `rsconnect.error.trace`. This option, when enabled, installs a custom error handler that prints stack traces when a local error occurs during deployment.  It's designed mostly to be used by the above switch, but can also be used standalone. 

Errors from **Packrat** are now explicitly marked as such, even in non-verbose mode. 

There is a planned IDE side of this change: We'll add an affordance to the IDE, probably in the publish preferences pane, to make it possible for customers (and developers) to turn on the verbose output mode for deployments originating from the IDE so they can share the diagnostic logs with the support team when that's necessary (hopefully rarely).

### CODE REVIEW

File: R/bundle.R [hunk 1]
```diff
@@ -470,7 +470,22 @@ addPackratSnapshot <- function(bundleDir, implicit_dependencies = c()) {
...
   # generate the packrat snapshot
-  performPackratSnapshot(bundleDir)
+  tryCatch({
+    performPackratSnapshot(bundleDir)
+  }, error = function(e) {
+    # if an error occurs while generating the snapshot, add a header to the
+    # message for improved attribution
+    e$msg <- paste0("----- Error snapshotting dependencies (Packrat) -----\n",
+                    e$msg)
+
```
Comment (ajmcoqui) [inline R/bundle.R]: Awesome.

File: R/deployApp.R [hunk 2]
```diff
@@ -85,13 +87,54 @@ deployApp <- function(appDir = getwd(),
...
-                      quiet = FALSE,
+                      logLevel = c("normal", "quiet", "verbose"),
                       lint = TRUE,
                       metadata = list()) {
 
   if (!isStringParam(appDir))
     stop(stringParamErrorMessage("appDir"))
 
+  # respect log level
+  logLevel <- match.arg(logLevel)
```
Comment (ajmcoqui) [inline R/deployApp.R]: I assume "normal" is the default logLevel; is that set explicitly somewhere, or is it de facto?

**Comment by jjallaire** [comment]
I'm assuming this is going into patch so we shouldn't merge it yet? (I haven't yet submitted rsconnect to CRAN, waiting to make sure the notebook announcement this week didn't turn up more bugs).

File: R/deployApp.R (hunk 2 above)
Comment (ajmcoqui) [inline R/deployApp.R]: Ah, TIL.  Thanks.

**Comment by jmcphers** [comment]
I'm planning on putting the IDE side of the change into patch, but this change can be merged at any time since it works with the IDE as-is.

**Comment by jjallaire** [comment]
Okay, let's wait until after the CRAN release then.

J

**Comment by jmcphers** [comment]
Okay, merging as we're past release + CRAN checkpoints.

**Pull request merged**